python/ray/data/datasource/torch_datasource.py

```diff
@@ -0,0 +1,91 @@
+import math
+from typing import TYPE_CHECKING
+
+import cloudpickle
+
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data.block import BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask
+
+if TYPE_CHECKING:
+    import torch
+
+
+class TorchDatasource(Datasource):
+    def create_reader(
+        self,
+        dataset: "torch.utils.data.Dataset",
+        batch_size: int = 32,
+        random_split: bool = False,
+    ):
+        return _TorchDatasourceReader(dataset, batch_size, random_split)
+
+
+class _TorchDatasourceReader(Reader):
+    def __init__(
+        self, dataset: "torch.utils.data.Dataset", batch_size: int, random_split: bool
+    ):
+        self._dataset = dataset
+        self._batch_size = batch_size
+        self._random_split = random_split
+        self._size = len(cloudpickle.dumps(self._dataset))
```

Comment:
i'm not sure if this is the best way to calculate dataset size, but i will let others comment since i'm not aware of other size calculation methods for torch Datasets.

Reply:
+1... I don't think there's a good way to calculate dataset size right now, just as with huggingface datasource